extensions/amp-ad-network-doubleclick-impl/0.1/test/test-amp-ad-network-doubleclick-impl.js

```diff
@@ -450,4 +491,355 @@ describes.sandboxed('amp-ad-network-doubleclick-impl', {}, () => {
+        });
+      if (opt_xhrFail) {
+        xhrWithArgs.returns(Promise.reject(
+            new TypeError('some random network error')));
+      } else if (opt_allInvalid) {
+        xhrWithArgs.throws(new Error('invalid should not make xhr!'));
+      } else {
+        xhrWithArgs.returns(Promise.resolve({
+          arrayBuffer: () => { throw new Error('Expected SRA!'); },
+          bodyUsed: false,
+          text: () => {
+            let slotDataString = '';
+            responses.forEach(slot => {
+              slotDataString +=
+                `${JSON.stringify(slot.headers)}\n${slot.creative}\n`;
+            });
+            return Promise.resolve(slotDataString);
+          },
+          headers: new FetchResponseHeaders({
+            getResponseHeader(name) {
+              return headers[name];
+            },
+          }),
+        }));
+      }
+    }
+
+    function generateNonSraXhrMockCall(impl, creative) {
+      // Start with nameframe method, SRA will override to use safeframe.
+      const headers = {};
+      headers[RENDERING_TYPE_HEADER] = XORIGIN_MODE.NAMEFRAME;
+      const iu = encodeURIComponent(impl.element.getAttribute('data-slot'));
+      const urlRegexp = new RegExp(
+        '^https:\/\/securepubads\\.g\\.doubleclick\\.net' +
+        `\/gampad\/ads\\?iu=${iu}&`);
+      xhrMock.withArgs(
+        sinon.match(urlRegexp),
+        {
+          mode: 'cors',
+          method: 'GET',
+          credentials: 'include',
+        }).returns(Promise.resolve({
+          arrayBuffer: () => utf8Encode(creative),
+          bodyUsed: false,
+          headers: new FetchResponseHeaders({
+            getResponseHeader(name) {
+              return headers[name];
+            },
+          }),
+          text: () => {
+            throw new Error('should not be SRA!');
+          },
+        }));
+    }
+
+    /**
+     * @param {!Array<number|{{
+     *    networkId:number,
+     *    instances:number,
+     *    xhrFail:boolean|undefined,
+     *    invalidInstances:number}}>} networkIds
+     */
+    function executeTest(items) {
+      const xhrFailingNetworks = {};
+      const allInvalidNetworks = {};
+      const instances = [];
+      items.forEach(network => {
+        if (typeof network == 'number') {
+          network = {networkId: network, instances: 1};
+        }
+        dev().assert(network.instances || network.invalidInstances);
+        const createInstances = (instanceCount, invalid) => {
+          for (let i = 0; i < instanceCount; i++) {
+            const impl = createInstance(network.networkId);
+            instances.push(impl);
+            if (invalid) {
+              sandbox.stub(impl, 'isValidElement').returns(() => false);
+              impl.element.setAttribute('data-test-invalid', 'true');
+            }
+          }
+        };
+        createInstances(network.instances);
+        createInstances(network.invalidInstances, true);
+        allInvalidNetworks[network.networkId] =
```

Comment:
this name isn't really right, seeing as all networks are included in this regardless of whether or not they are invalid. Maybe something like `networkValidity`?